File: shared/types/stock.ts
export interface OHLCVData {
  time: Date;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  symbol?: string;
  frequency?: Frequency;
}

export interface OHLCData {
  time: Date;
  open: number;
  high: number;
  low: number;
  close: number;
}

export interface PriceData {
  value: number;
  time: Date;
}

export interface VolumeData {
  value: number;
  time: Date;
  color: string;
}
export interface Ticker {
  symbol: string;
  name?: string;
  exchange?: string;
}

export interface LiveTickerFeedMessage {
  dailyBar: OHLCVData | null;
  minuteBar: OHLCVData | null;
}

export interface StockBasicFinancials {
  // Valuation
  peTTM?: number;
  peAnnual?: number;
  forwardPE?: number;
  pb?: number;
  priceToSalesTTM?: number;
  pegTTM?: number;
  evToEBITDA?: number;
  evToRevenue?: number;
  enterpriseValue?: number;

  // Earnings
  epsTTM?: number;
  epsAnnual?: number;
  forwardEps?: number;

  // Profitability
  roe?: number;
  roa?: number;
  roic?: number;
  netProfitMargin?: number;
  grossMargin?: number;
  operatingMargin?: number;

  // Growth
  revenueGrowth?: number;
  epsGrowth?: number;

  // Cash Flow
  freeCashFlow?: number;
  operatingCashFlow?: number;

  // Financial Health
  debtToEquity?: number;
  totalDebt?: number;
  totalCash?: number;
  netDebt?: number;
  currentRatio?: number;
  quickRatio?: number;
  interestCoverage?: number;

  // Per Share
  bookValue?: number;
  revenuePerShare?: number;
  dividendPerShare?: number;

  // Dividends
  divYieldTTM?: number;
  payoutRatio?: number;

  // Market
  marketCap?: number;
  sharesOutstanding?: number;
  yearHigh?: number;
  yearLow?: number;

  // Volume
  adtv10Day?: number;

  // Risk
  beta?: number;
}

export type Frequency = "intraday" | "daily" | "weekly" | "monthly";

File: backend/app/schemas/stock.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

type Frequency = Literal["intraday", "daily", "weekly", "monthly"]

class OHLCVData(BaseModel):
    time: datetime
    open: float
    high: float
    low: float
    close: float
    volume: float
    symbol: str | None = None
    frequency: Frequency | None = None


class OHLCData(BaseModel):
    time: datetime
    open: float
    high: float
    low: float
    close: float


class PriceData(BaseModel):
    value: float
    time: datetime


class VolumeData(BaseModel):
    value: float
    time: datetime
    color: str


class Ticker(BaseModel):
    symbol: str
    name: str | None = None
    exchange: str | None = None


class LiveTickerFeedMessage(BaseModel):
    dailyBar: OHLCVData | None = None
    minuteBar: OHLCVData | None = None


class StockBasicFinancials(BaseModel):
    # Valuation
    peTTM: float | None = None
    forwardPE: float | None = None
    pb: float | None = None
    priceToSalesTTM: float | None = None
    pegTTM: float | None = None
    evToEBITDA: float | None = None
    evToRevenue: float | None = None
    marketCap: float | None = None
    enterpriseValue: float | None = None
    # Earnings
    epsTTM: float | None = None
    forwardEps: float | None = None  # not available in Finnhub metric response
    # Profitability
    roe: float | None = None
    netProfitMargin: float | None = None
    grossMargin: float | None = None
    operatingMargin: float | None = None
    # Growth
    revenueGrowth: float | None = None
    epsGrowth: float | None = None
    # Financial Health
    debtToEquity: float | None = None
    totalDebt: float | None = None  # not available in Finnhub metric response
    totalCash: float | None = None  # not available in Finnhub metric response
    netDebt: float | None = None  # requires totalDebt and totalCash
    currentRatio: float | None = None
    quickRatio: float | None = None
    interestCoverage: float | None = None
    # Per Share
    bookValue: float | None = None
    revenuePerShare: float | None = None
    dividendPerShare: float | None = None
    # Dividends
    divYieldTTM: float | None = None
    payoutRatio: float | None = None
    # Cash Flow
    freeCashFlow: float | None = None  # Finnhub only provides per-share values
    operatingCashFlow: float | None = None  # Finnhub only provides per-share values
    # Market
    sharesOutstanding: float | None = None  # not available in Finnhub metric response
    yearHigh: float | None = None
    yearLow: float | None = None
    # Volume
    adtv10Day: float | None = None
    # Risk
    beta: float | None = None
